drivers/adc/adc_gd32.c

```diff
@@ -0,0 +1,461 @@
+	ADC_RSQ2(cfg->reg) &= ~ADC_RSQX_RSQN;
+	ADC_RSQ2(cfg->reg) = index;
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_gd32_read(const struct device *dev,
+			 const struct adc_sequence *sequence)
+{
+	struct adc_gd32_data *data = dev->data;
+	int error;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	error = adc_gd32_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_gd32_read_async(const struct device *dev,
+			       const struct adc_sequence *sequence,
+			       struct k_poll_signal *async)
+{
+	struct adc_gd32_data *data = dev->data;
+	int error;
+
+	adc_context_lock(&data->ctx, true, async);
+	error = adc_gd32_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+#endif /* CONFIG_ADC_ASYNC */
+
+static struct adc_driver_api adc_gd32_driver_api = {
+	.channel_setup = adc_gd32_channel_setup,
+	.read = adc_gd32_read,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = adc_gd32_read_async,
+#endif /* CONFIG_ADC_ASYNC */
+};
+
+static int adc_gd32_init(const struct device *dev)
+{
+	struct adc_gd32_data *data = dev->data;
+	const struct adc_gd32_config *cfg = dev->config;
+	int ret;
+
+	data->dev = dev;
+
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+#ifdef CONFIG_SOC_SERIES_GD32F3X0
+	/* Select adc clock source and its prescaler. */
+	rcu_adc_clock_config(cfg->rcu_clock_source);
+#endif
+
+	rcu_periph_clock_enable(cfg->rcu_periph_clock);
+
+#if defined(CONFIG_SOC_SERIES_GD32F403) || \
+	defined(CONFIG_SOC_SERIES_GD32VF103) || \
+	defined(CONFIG_SOC_SERIES_GD32F3X0)
+	/* Set SWRCST as the regular channel external trigger. */
+	ADC_CTL1(cfg->reg) &= ~ADC_CTL1_ETSRC;
+	ADC_CTL1(cfg->reg) |= CTL1_ETSRC(7);
+
+	/* Enable external trigger for regular channel. */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ETERC;
+#endif
+
+	/* Enable ADC */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ADCON;
+
+	adc_gd32_calibration(cfg);
+
+	cfg->irq_config_func();
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+#define HANDLE_SHARED_IRQ(n, active_irq)							\
+	static const struct device *dev_##n = DEVICE_DT_INST_GET(n);				\
+	const struct adc_gd32_config *cfg_##n = dev_##n->config;				\
+												\
+	if ((cfg_##n->irq_num == active_irq) &&							\
+		(ADC_CTL0(cfg_##n->reg) & ADC_CTL0_EOCIE)) {					\
+		adc_gd32_isr(dev_##n);								\
+	}
+
+static void adc_gd32_global_irq_handler(const struct device *dev)
+{
+	const struct adc_gd32_config *cfg = dev->config;
+
+	LOG_DBG("global irq handler: %u", cfg->irq_num);
+
+	DT_INST_FOREACH_STATUS_OKAY_VARGS(HANDLE_SHARED_IRQ, (cfg->irq_num));
+}
+
+static void adc_gd32_global_irq_cfg(void)
+{
+	static bool global_irq_init = true;
+
+	if (!global_irq_init) {
+		return;
+	}
+
+	global_irq_init = false;
```

Comment:
Why not use positive logic?

```c
	static bool global_irq_init = false;

	if (global_irq_init) {
		return;
	}

	global_irq_init = true;
```

Reply:
done.

Reply:
An error occured, `ERROR:INITIALISED_STATIC: do not initialise statics to false`.
Revert to positive init value.